modules/SequencesExt.tla

```diff
@@ -124,6 +124,37 @@ Last(s) ==
 
 -----------------------------------------------------------------------------
 
+IsSubSequence(s, t) ==
+  (**************************************************************************)
+  (* TRUE iff the sequence s is a subsequence of the sequence t, s.t.       *)
+  (**************************************************************************)
+  \E i,j \in 1..Len(t):
+    /\ i <= j \* Optimization to not eval SubSeq with j < i, which is defined as <<>>.
+    /\ SubSeq(t, i, j) = s
+
+ReplaceFirstSubSeq(r, s, t) ==
+  (**************************************************************************)
+  (* The sequence  t  with its subsequence  s  replaced by the sequence  r  *)
+  (**************************************************************************)
+  IF IsSubSequence(s, t) THEN \* Prevent CHOOSE x \in S: P not holding for any element.
+    LET idx == CHOOSE i \in 1..Len(t): \E j \in i..Len(t): SubSeq(t, i, j) = s
```

Comment:
The name of the operator suggests that the first occurrence of s in t will be replaced by r, but the definition doesn't enforce this. One could either rename the operator or change the definition to

LET idx == CHOOSE i \in 1 .. Len(t) : 
                        /\ SubSeq(t, i, Len(s)) = s
                        /\ \A j \in 1 .. i-1 : SubSeq(t, j, Len(s)) # s

I presume that the way TLC implements CHOOSE ensures that the first match will be chosen, though.